test/e2e/clusterctl_upgrade.go

```diff
@@ -69,11 +69,13 @@ type ClusterctlUpgradeSpecInput struct {
 	// provider contract to use to initialise the secondary management cluster, e.g. `v1alpha3`
 	InitWithProvidersContract string
 	SkipCleanup               bool
+	PreCreateCluster          func(managementClusterProxy framework.ClusterProxy, managementClusterNamespace string)
```

Comment:
Additionally I would suggest to also hand over the cluster name (`managementClusterName`) so the func gets the Cluster namespace+name

Reply:
(Assuming it's fine for your use case that the func is called after we apply the cluster resources and before we wait for the cluster to be provisioned. If it's not we can't use the `PreWaitForCluster` func of `ApplyClusterTemplateAndWait`)

Reply:
Yeah I think It would be easier just use `PreWaitForCluster` and propagate it to the provider from here and also I will move it in the second call after ```Expect(managementClusterProxy.Apply(ctx, workloadClusterTemplate)).To(Succeed())```